### DISCUSSION TITLE
Manually specify sourceUrl for a Docker package

### DISCUSSION THREAD

**Discussion by insertjokehere** [renovatebot/renovate#7947] [Request Help] [answered]
### Which Renovate are you using?

Renovate Open Source CLI

### Which platform are you using?

Gitea self-hosted

### Have you checked the logs? Don't forget to include them if relevant

```
 INFO: Found renovate config errors (repository=kubernetes/helm-configs)
       "errors": [
         {
           "depName": "Configuration Error",
           "message": "Invalid configuration option: packageRules[3].sourceUrl"
         }
       ]
 INFO: Repository has invalid config (repository=kubernetes/helm-configs)
       "error": {
         "configFile": "renovate.json",
         "validationError": "The renovate configuration file contains some invalid settings",
         "validationMessage": "Invalid configuration option: packageRules[3].sourceUrl",
         "message": "config-validation",
         "stack": "Error: config-validation\n    at mergeRenovateConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/init/config.ts:164:19)\n    at Object.getRepoConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/init/config.ts:238:12)\n    at Object.initRepo (/usr/src/app/node_modules/renovate/lib/workers/repository/init/index.ts:21:12)\n    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:37:14)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:67:7)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:10:22"
       }
```

### What would you like to do?

I'm trying to manually specify a source URL for a Docker package that doesn't include a source label so I get proper change logs. Following advice from https://github.com/renovatebot/renovate/issues/6814#issuecomment-741702288, I added a package rule that specifies a `sourceUrl` property, but this is rejected as invalid. Am I doing something wrong, or am I misinformed?

Full `renovate.json` file:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "helm-values": {
    "fileMatch": ["values\\.yaml$"],
    "ignorePaths": ["charts/"]
  },
  "helm-requirements": {
    "enabled": false
  },
  "helmv3": {
    "enabled": false
  },
  "packageRules": [
    {
      "packageNames": ["lmierzwa/karma", "quay.io/wekan/wekan"],
      "versioning": "loose"
    },
    {
      "packageNames": ["nextcloud"],
      "versioning": "docker"
    },
    {
      "packageNames": ["lmierzwa/karma"],
      "allowedVersions": "/^v[0-9]+\\.[0-9]+$/"
    },
    {
      "packageNames": ["quay.io/wekan/wekan"],
      "sourceUrl": "https://github.com/wekan/wekan"
    }
  ],
  "regexManagers": [
    {
      "fileMatch": ["^build_config\\.yaml$"],
      "matchStrings": [
        "datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\s.*_version: (?<currentValue>.*)\\s"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    },
    {
      "fileMatch": ["(^|/)releases/[^/]+\\.yaml$"],
      "matchStrings": [
        "datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?( registryUrl=(?<registryUrl>.*?))?\\s.*version: (?<currentValue>.*)\\s"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    }
  ]
}
```

**Comment by viceice** [comment]
The workaround is to ad the sourceUrl package rule to the bot config.js. as it's not validated there. But applied. 😏

**Reply by insertjokehere** [reply]
Thanks, this lets me do what I want. For future people looking at this, adding something like:
```
 "packageRules": [
     {
       "packageNames": ["quay.io/wekan/wekan"],
       "sourceUrl": "https://github.com/wekan/wekan"
     }
]
```
To the `config.json` file provided to the self-hosted bot works as intended (although is undocumented, hopefully this doesn't break in the future?)

**Comment by rarkins** [comment]
FYI one reason to not allow sourceUrl to be repository-configurable currently is because it could totally mess with our caching if someone set a public package to have wrong sourceUrl. We have this code for any *public* sourceUrl additions: https://github.com/renovatebot/renovate/blob/e18259c2a5cb6ddc1d8a7939aeeace86c0d2c5a7/lib/datasource/metadata.ts#L61

**Comment by rarkins** [comment, marked as answer]
Summarizing in order of preference:

1. Docker images ideally include the source repo in their metadata, e.g. with `LABEL org.opencontainers.image.source="https://github.com/renovatebot/renovate"`
2. If it's a public image missing source URL, contribute it to this repo here: https://github.com/renovatebot/renovate/blob/e18259c2a5cb6ddc1d8a7939aeeace86c0d2c5a7/lib/datasource/metadata.ts#L61
3. If it's a private image, add it via `packageRules` in the *bot* config
